KAFKA-13830 MetadataVersion integration for KRaft controller

mumrah opened the pull request:
This patch builds on #12072 and adds controller support for `metadata.version`. The kafka-storage tool now allows a user to specify a specific metadata.version to bootstrap into the cluster, otherwise the latest version is used.

Upon the first leader election of the KRaft quroum, this initial metadata.version is written into the metadata log. When writing snapshots, a FeatureLevelRecord for metadata.version will be written out ahead of other records so we can decode things at the correct version level.

This also includes additional validation in the controller when setting feature levels. It will now check that a given metadata.version is supportable by the quroum, not just the brokers.

mumrah committed:
KAFKA-13830 Introduce metadata.version

This patch includes a new metadata.version which is planned to replace IBP in
KRaft clusters. The kafka-storage tool now allows a user to specify a specific
metadata.version to bootstrap into the cluster, otherwise the latest version
is used.

Upon the first leader election of the KRaft quroum, this initial metadata.version
is written into the metadata log. When writing snapshots, a FeatureLevelRecord
for metadata.version will be written out ahead of other records so we can decode
things at the correct version level.

This also includes additional validation in the controller when setting feature
levels. It will now check that a given metadata.version is supportable by the
quroum, not just the brokers.

mumrah committed:
Some cleanup after splitting the PR

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -883,15 +905,49 @@ public void handleLeaderChange(LeaderAndEpoch newLeader) {
+                                    log.warn("Renouncing the leadership at oldEpoch {} since we could not bootstrap" +
+                                             "a metadata.version. Reverting to last committed offset {}.",
+                                        curClaimEpoch, lastCommittedOffset);
+                                    renounce();
```
@cmccabe is this the correct way to renounce leadership? Originally, I was calling renounce directly here, but that led to a ConcurrentModificationException

dengziming commented on core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java:
```diff
@@ -237,6 +239,15 @@ public SocketServer anyControllerSocketServer() {
+                KafkaServer::brokerFeatures
+            ));
+        }
+
```
nit: redundant inline

dengziming commented on core/src/main/scala/kafka/server/BrokerMetadataCheckpoint.scala:
```diff
@@ -102,11 +112,16 @@ class RawMetaProperties(val props: Properties = new Properties()) {
     val clusterId = require(ClusterIdKey, properties.clusterId)
     val nodeId = require(NodeIdKey, properties.nodeId)
-    new MetaProperties(clusterId, nodeId)
+    val metadataVersion = properties.initialMetadataVersion.getOrElse(MetadataVersion.stable().version())
```
Should this be MetadataVersion.V1? is this the case of the metadata version before 3.2.0?

mumrah replied:
You're right, this should be V1. This covers the case of an existing meta.properties file that does not have the initial metadata.version specified. We're using this as a signal that a cluster was on a KRaft preview version previously.

dengziming reviewed (commented):
Thank you for this PR @mumrah , I have a question: should we bump IBP from 3.2-0 to 3.2-1, then in the future when migrating ZooKeeper clusters to KRaft, we can set metadataVersion=1 if the ibp is behind 3.2-0, and set metadataVersion=2 if ibp equals 3.2-1.

mumrah committed:
PR feedback

mumrah committed:
Merge remote-tracking branch 'apache-github/trunk' into KAFKA-13830-introduce-metadata-version

mumrah committed:
Fix compile failure after merge from trunk

mumrah committed:
Try reducing log clearner buffer size

mumrah committed:
Try increasing heap size for integration tests

mumrah committed:
Reduce increased heap to 2.2g

mumrah committed:
Fix invalid heap size value

mumrah committed:
Trying something for Jenkins

mumrah committed:
Merge remote-tracking branch 'apache-github/trunk' into KAFKA-13830-introduce-metadata-version

mumrah committed:
Workaround compiler warning from trunk

mumrah committed:
Merge remote-tracking branch 'apache-github/trunk' into KAFKA-13830-introduce-metadata-version

mumrah committed:
Revert "Workaround compiler warning from trunk"

This reverts commit 6002e817b587b2ca7f52a9c8f4c39d48dd46f5f2.

mumrah committed:
Bump to 2.5g heap

mumrah commented:
@dengziming @cmccabe FYI, there is another PR coming soon which includes a big consolidation of ApiVersion and MetadataVersion. I will wait to merge this one until that one is done.

mumrah committed:
Update with MetadataVersion changes

mumrah committed:
Fixup method names before merge from trunk

mumrah committed:
Merge remote-tracking branch 'apache-github/trunk' into KAFKA-13830-introduce-metadata-version

mumrah committed:
post merge fixup

mumrah commented:
@dengziming @cmccabe I've updated this PR with the changes from #12072. I've also modified the bootstrap mechanism to use a file with the KRaft snapshot format rather than a property in meta.properties. This is a slight departure from the KIP, but using a snapshot will open up the ability for us to bootstrap other records on first startup. If we agree that this is a good approach, I can update the KIP and discussion thread.

dengziming commented on clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java:
```diff
@@ -299,4 +299,30 @@ public static ApiVersion toApiVersion(ApiKeys apiKey) {
             .setMaxVersion(apiKey.latestVersion());
     }
+
+    public static ApiVersionsResponse apiVersionsResponse(
```
This method is identical to another `createApiVersionsResponse` method, so we can remove it

dengziming commented on core/src/main/scala/kafka/server/BrokerMetadataCheckpoint.scala:
```diff
@@ -35,6 +35,7 @@ object RawMetaProperties {
   val BrokerIdKey = "broker.id"
   val NodeIdKey = "node.id"
   val VersionKey = "version"
+  val InitialMetadataVersion = "init.metadata.version"
```
It seems this variable is not used.

dengziming commented on core/src/main/scala/kafka/server/KafkaRaftServer.scala:
```diff
@@ -102,6 +106,8 @@ class KafkaRaftServer(
       controllerQuorumVotersFuture,
       KafkaRaftServer.configSchema,
+      raftApiVersions,
+      Some(bootstrapMetadata)
```
I have 1 concern:
we should also reset featureLevel of `BrokerServer.FinalizedFeatureCache`?  I think featureLevel is similar to ibp of ZK broker, they should be the same as possible.

mumrah replied:
That's an interesting point. For the broker, on first start, it will start up without a `metadata.version`. It won't be until the bootstrapped version is written by the controller and replicated to the broker that BrokerMetadataPublisher will update the FinalizedFeatureCache. 

This is probably okay though, since brokers start up in a fenced state in KRaft anyways. Since the `metadata.version` bootstrapping is done in the leader election callback on the controller, we should have the FeatureLevelRecord written out before any broker registrations occur. I think this should avoid any issues with a broker becoming unfenced before knowing what the `metadata.version` is.

Does that make sense? Or did I misunderstand your question :)

dengziming replied:
Yeah, that makes sense. But I want to clarify one thing here:
> we should have the FeatureLevelRecord written out before any broker registrations occur

The broker will register itself regardless of whether FeatureLevelRecord is written, but only set wantfence=false in the HeartbeatReq after FeatureLevelRecord has been published.

dengziming reviewed (commented):
I haven't gone through all the code, just left some questions firstly.

dengziming commented on metadata/src/main/java/org/apache/kafka/image/FeaturesDelta.java:
```diff
@@ -51,7 +58,11 @@ public void finishSnapshot() {
+        }
     }
 
     public void replay(RemoveFeatureLevelRecord record) {
```
Even though we didn't have `RemoveFeatureLevelRecord` currently, should we also set metadataVersionChange = Empty?

dengziming commented on metadata/src/main/java/org/apache/kafka/controller/QuorumFeatures.java:
```diff
@@ -17,34 +17,76 @@
+            ));
+        }
+    }
+
```
nit: redudant inline

mumrah committed:
Merge remote-tracking branch 'apache-github/trunk' into KAFKA-13830-introduce-metadata-version

mumrah commented:
@dengziming addressing your old comment:

> Thank you for this PR @mumrah , I have a question: should we bump IBP from 3.2-0 to 3.2-1, then in the future when migrating ZooKeeper clusters to KRaft, we can set metadataVersion=1 if the ibp is behind 3.2-0, and set metadataVersion=2 if ibp equals 3.2-1.

With #12072, we have associated a `metadata.version` with each IBP starting with 3.0-IV0. In this PR, I've added 3.3-IV0 which will be metadata version 5.

mumrah committed:
PR feedback

mumrah committed:
Add a test for BootstrapMetadata

ahuang98 commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -91,6 +97,23 @@ public static NodeApiVersions create(short apiKey, short minVersion, short maxVe
     }
 
+    public NodeApiVersions(ApiVersionCollection nodeApiVersions, SupportedFeatureKeyCollection nodeSupportedFeatures) {
+        for (ApiVersion nodeApiVersion : nodeApiVersions) {
```
Can we call `NodeApiVersions(ApiVersionCollection nodeApiVersions)` here to reduce some redundant code?

mumrah commented on core/src/main/scala/kafka/server/ControllerServer.scala:
```diff
@@ -63,6 +64,8 @@ class ControllerServer(
   val controllerQuorumVotersFuture: CompletableFuture[util.Map[Integer, AddressSpec]],
   val configSchema: KafkaConfigSchema,
+  val raftApiVersions: ApiVersions,
+  val bootstrapMetadata: Option[BootstrapMetadata] = None
```
I still need to remove this Option and make it a regular argument.

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -55,25 +60,37 @@ public class FeatureControlManager {
         this.log = logContext.logger(FeatureControlManager.class);
         this.quorumFeatures = quorumFeatures;
         this.finalizedVersions = new TimelineHashMap<>(snapshotRegistry, 0);
+        this.metadataVersion = new TimelineObject<>(snapshotRegistry, MetadataVersion.UNINITIALIZED);
```
UNINITIALIZED was added since we need a non-null object to initialize the TimelineObject

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -233,4 +256,8 @@ public ApiVersion apiVersion(ApiKeys apiKey) {
     }
+
+    public Map<String, SupportedVersionRange> supportedFeatures() {
+        return Collections.unmodifiableMap(supportedFeatures);
```
It seems like `supportedFeatures` should be immutable and final in any case. There is no use-case for mutating this after the object is constructed.

mumrah committed:
Clean up a few things left from prior to the MetadataVersion PR

ahuang98 commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -273,6 +301,40 @@ public static MetadataVersion latest() {
+        }
+    }
+
+    public static boolean checkIfMetadataChanged(MetadataVersion sourceVersion, MetadataVersion targetVersion) {
```
Just clarifying, this is to check if the metadata has changed in an incompatible way between the sourceVersion and targetVersion?

ahuang98 replied:
Let's add a test for this

mumrah replied:
That's right, this will test if two versions have compatible metadata based on the `didMetadataChange` boolean added to the enum

ahuang98 commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -273,6 +301,40 @@ public static MetadataVersion latest() {
     }
 
+    public Optional<MetadataVersion> previous() {
+        int idx = this.ordinal();
```
The indexing for `ordinal()` starts at `0` right? With `MetadataVersion.UNINITIALIZED` at index 0 (which I'm assuming we don't want to count as a valid previous version), should the conditional be `if (idx > 1)`?

mumrah replied:
Good catch! I think this was left over from when I had put UNINITIALIZED into VALUES. I opted against that since it would be a gotcha for future developers to encounter a version that was not really a version.

ahuang98 commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -273,6 +301,40 @@ public static MetadataVersion latest() {
+    public Optional<MetadataVersion> previous() {
+        int idx = this.ordinal();
+        if (idx > 2) {
+            return Optional.of(MetadataVersion.values()[idx - 1]);
```
```suggestion
            return Optional.of(VALUES[idx]);
```
The indexing is a bit unintuitive, I wonder if you might just want to store all of `values()` (including `UNINITIALIZED`) in `VALUES` just to make this more clear?

mumrah replied:
I'm going to reorganize this code a bit to make it more clear.

ahuang98 commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -70,13 +47,17 @@ object StorageTool extends Logging {
           val clusterId = namespace.getString("cluster_id")
+          val metadataVersion = getMetadataVersion(namespace)
+          if (metadataVersion.isLessThan(MetadataVersion.IBP_3_0_IV0)) {
+            throw new TerseFailure(s"Cannot specify a metadata version less than ${MetadataVersion.IBP_3_0_IV0.featureLevel()}.")
```
nit: Create a `MetadataVersion.isKRaftSupported()` / `MetadataVersion.firstKRaftMetadataVersion()` method? You could use it in 
https://github.com/apache/kafka/pull/12050/files#diff-2ee48010b1035f21f5ebdd47a78fc632ed239179aee77e66d75a6bdb77ee8fd7R147 
and
https://github.com/apache/kafka/pull/12050/files#diff-602affccc2af320d6cdfda31afc197bf1a798497009c404706a600ab57c1bea5R78
to name a few places.

mumrah replied:
Good idea, I'll add this.

ahuang98 commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -189,21 +207,24 @@ object StorageTool extends Logging {
   def buildMetadataProperties(
     clusterIdStr: String,
-    config: KafkaConfig
+    config: KafkaConfig,
```
nit: extra comma?

ahuang98 commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:
```diff
@@ -127,19 +128,26 @@ class BrokerMetadataPublisher(conf: KafkaConfig,
       // Publish the new metadata image to the metadata cache.
       metadataCache.setImage(newImage)
 
+      val metadataVersion: Option[Short] = Option(newImage.features().metadataVersion())
```
Is it expected that we might end up with an empty Option here? Wondering if we need any error logging for this

ahuang98 commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:
```diff
@@ -127,19 +128,26 @@ class BrokerMetadataPublisher(conf: KafkaConfig,
+
       if (_firstPublish) {
-        info(s"Publishing initial metadata at offset $highestOffsetAndEpoch.")
+        info(s"Publishing initial metadata at offset $highestOffsetAndEpoch  with metadata.version $metadataVersion.")
```
nit: remove extra space before `with`

ahuang98 commented on core/src/test/scala/unit/kafka/server/BrokerFeaturesTest.scala:
```diff
@@ -25,10 +25,6 @@ import scala.jdk.CollectionConverters._
 
 class BrokerFeaturesTest {
 
-  @Test
```
nit: Instead of removing, should we replace with `assertTrue(BrokerFeatures.createEmpty().supportedFeatures.empty)`

ahuang98 commented on core/src/main/scala/kafka/raft/RaftManager.scala:
```diff
@@ -108,7 +108,8 @@ class KafkaRaftManager[T](
   threadNamePrefixOpt: Option[String],
-  val controllerQuorumVotersFuture: CompletableFuture[util.Map[Integer, AddressSpec]]
+  val controllerQuorumVotersFuture: CompletableFuture[util.Map[Integer, AddressSpec]],
+  apiVersions: ApiVersions
```
Not sure if it makes a difference, but it seems that all calls to create a new `KafkaRaftManager` use a newly created `ApiVersions`, should we just drop this from the list of constructor arguments and instantiate within the class instead? 
```
private val apiVersions = new ApiVersions()
```

mumrah replied:
This was done in order to expose KafkaRaftManager's ApiVersions to QuorumFeatures which is used by the controller. ApiVersions gets populated with NodeApiVersions which includes supported APIs and supported features. This lets us check if the quorum supports a given feature flag and feature version.

cmccabe replied:
I don't see why we would want to pass a specific `ApiVersions` object in. We're not supplying any information, right? We don't know anything about the versions of the nodes until we start contacting them.

So, can we just have an accessor method in `RaftManager` that returns the `ApiVersions` of the `RaftManager`?

mumrah committed:
More feedback from PR

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java:
```diff
@@ -0,0 +1,196 @@
+        void handleMessage(ApiMessage message) {
+            try {
+                MetadataRecordType type = MetadataRecordType.fromId(message.apiKey());
+                messages.computeIfAbsent(type, __ -> new ArrayList<>()).add(message);
```
Should we be adding the message to the map even if its `type` is already a key in the map?

mumrah replied:
The idea here was just to collect all messages into a multi-map. In the future we will probably put more records into the bootstrap snapshot.

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -882,16 +897,55 @@ public void handleLeaderChange(LeaderAndEpoch newLeader) {
+                            metadataVersion = MetadataVersion.UNINITIALIZED;
+                            future = new CompletableFuture<>();
+                            future.completeExceptionally(
+                                    new IllegalStateException("Cannot become leader without a valid initial metadata.version to use. Got " + bootstrapMetadataVersion.version()));
```
nit: Add the min viable MetadataVersion in the log.

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -83,6 +100,25 @@ ControllerResult<Map<String, ApiError>> updateFeatures(
+                Collections.singletonMap(
+                    MetadataVersion.FEATURE_NAME,
+                    new ApiError(Errors.INVALID_UPDATE_VERSION,
+                        "Cannot initialize metadata.version since it has already been initialized.")
```
nit: Add the metadata.version in use to the error message

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -93,84 +129,187 @@ boolean featureExists(String featureName) {
+    ) {
+        Optional<VersionRange> quorumSupported = quorumFeatures.quorumSupportedFeature(MetadataVersion.FEATURE_NAME);
+        if (!quorumSupported.isPresent()) {
+            return invalidMetadataVersion(newVersionLevel, "The quorum does not support metadata.version.");
```
This is making me realize it might be confusing that `metadata.version` refers to `featureLevel` when we also have `MetadataVersion` the enum.

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -93,84 +129,187 @@ boolean featureExists(String featureName) {
+        }
+
+        if (newVersionLevel <= 0) {
+            return invalidMetadataVersion(newVersionLevel, "metadata.version cannot be less than 1.");
```
nit
```suggestion
            return invalidMetadataVersion(newVersionLevel, "KRaft mode/the quorum does not support metadata.version values less than 1.");
```

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -93,84 +129,187 @@ boolean featureExists(String featureName) {
+     */
+    private ApiError updateMetadataVersion(
+        short newVersionLevel,
+        boolean allowUnsafeDowngrade,
```
Are you planning on supporting `allowUnsafeDowngrade == true` in this PR?

mumrah replied:
No, this will be done in a future PR. We will also need the code to generate a snapshot at the downgraded level. I filed KAFKA-13410 to track this

dengziming reviewed (commented):
Thank you for your reply, left 2 minor improvements.

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -93,84 +129,187 @@ boolean featureExists(String featureName) {
+            log.info("Setting feature {} to {}", record.name(), record.featureLevel());
+            finalizedVersions.put(record.name(), record.featureLevel());
+        }
+        for (Entry<String, FeatureLevelListener> listenerEntry : listeners.entrySet()) {
```
What other FeatureLevelListeners might we have (in addition to `QuorumFeatureListener`)?

mumrah replied:
I can't really think of any immediate use cases. It's possible we'll just have a single listener on the controller side.

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -93,84 +129,187 @@ boolean featureExists(String featureName) {
     class FeatureControlIterator implements Iterator<List<ApiMessageAndVersion>> {
         private final Iterator<Entry<String, Short>> iterator;
+        private final MetadataVersion metadataVersion;
+        private boolean wroteVersion = false;
```
Would it work to name this something like `isUninitializedVersion` or simply `uninitialized`? The point of this var is just to avoid doing the `metadataVersion.equals(UNINITIALIZED)` comparison every time we call `FeatureControlIterator.next()` right?

mumrah replied:
This boolean is used to make sure we write out the metadata.version feature record before any others. If (somehow) a snapshot is generated before the metadata.version is initialized, we skip it (hence the check a few lines below this)

ahuang98 commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -927,6 +981,31 @@ private void appendRaftEvent(String name, Runnable runnable) {
+                    }
+                    beginShutdown();
+                }
+            }
```
Is the behavior dependent on the feature being metadata.version?
```suggestion
            if (!isFeatureSupported) {
                if (isActiveController) {
                    log.error("Active controller cannot support {} {}", featureName, finalizedVersion);
                } else {
                    log.error("Standby controller cannot support {} {}, shutting down.", featureName, finalizedVersion);
                }
                beginShutdown();
            }
```

mumrah replied:
Yea, that's a good point. If the controller receives a record for a feature it can't support it should probably shutdown.

mumrah committed:
Update metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

Co-authored-by: Alyssa Huang <ahuang@confluent.io>

mumrah committed:
Update metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

Co-authored-by: Alyssa Huang <ahuang@confluent.io>

mumrah committed:
Address PR review comments

mumrah committed:
Fix a KRaftClusterTest

mumrah committed:
Merge remote-tracking branch 'mumrah/KAFKA-13830-introduce-metadata-version' into KAFKA-13830-introduce-metadata-version

mumrah committed:
Write all records from boostrap.snapshot into log on first start

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java:
```diff
@@ -0,0 +1,198 @@
+public class BootstrapMetadata {
+    private static final Logger log = LoggerFactory.getLogger(BootstrapMetadata.class);
+
+    public static final String BOOTSTRAP_FILE = "bootstrap.snapshot";
```
Can we use a different suffix? Metadata snapshots use `.checkpoint`. We did this because Kafka already uses `snapshot` for producer snapshots.

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java:
```diff
@@ -0,0 +1,198 @@
+                return type.equals(MetadataRecordType.FEATURE_LEVEL_RECORD);
+            })
+            .map(message -> (FeatureLevelRecord) message.message())
+            .filter(record -> record.name().equals(MetadataVersion.FEATURE_NAME))
```
Minor but a `flatMap` may be easier to read. For example this pseudo code:
```java
            .flatMap(message -> {
                MetadataRecordType type = MetadataRecordType.fromId(message.message().apiKey());
                if (!type.equals(MetadataRecordType.FEATURE_LEVEL_RECORD)) return Stream.empty();

                FeateureLevelRecord record = message.message();
                if (!record.name().equals(MetadataVersion.FEATURE_NAME)) return Stream.empty();

                return Stream.of(record);
            })
```

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java:
```diff
@@ -0,0 +1,198 @@
+            throw new IOException("Cannot write metadata bootstrap file " + bootstrapPath +
+                ". File already already exists.");
+        }
+        SnapshotFileWriter bootstrapWriter = SnapshotFileWriter.open(bootstrapPath);
+        bootstrapWriter.append(metadata.records());
+        bootstrapWriter.close();
```
You can use Java's "try with resource".

mumrah replied:
Thanks, I always forget about that :)

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java:
```diff
@@ -0,0 +1,198 @@
+
+        BootstrapListener listener = new BootstrapListener();
+        SnapshotFileReader reader = new SnapshotFileReader(bootstrapPath.toString(), listener);
+        reader.startup();
```
I think this starts a thread that needs to be shutdown.

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java:
```diff
@@ -0,0 +1,198 @@
+        SnapshotFileReader reader = new SnapshotFileReader(bootstrapPath.toString(), listener);
+        reader.startup();
+        try {
+            reader.caughtUpFuture().get();
```
It looks like this implementation starts a thread and blocks on that thread reading the bootstrap snapshot? Why is the thread needed? Why not read the snapshot on this thread?

mumrah replied:
I am reusing the snapshot reader that @cmccabe added for the metadata shell which uses a queue and thread. I agree that a thread isn't necessary here, but I figured the code reuse was better than adding yet another a snapshot reader

cmccabe replied:
Yeah. The thread was useful for the metadata shell since the idea was it would keep handling new updates in the background. Obviously, that isn't needed here, but if we're reusing the code, we'll get it.

Long-term, we could have a version of EventQueue that uses the current thread (kind of like Java's executor that does the same thing, can't remember what it's called) and use it here. It wouldn't be hard to write

jsancio reviewed (commented):
Thanks for the chances. Partial review.

cmccabe commented on core/src/main/scala/kafka/server/BrokerFeatures.scala:
```diff
@@ -72,6 +73,12 @@ class BrokerFeatures private (@volatile var supportedFeatures: Features[Supporte
 
   def createDefault(): BrokerFeatures = {
+    new BrokerFeatures(Features.supportedFeatures(
+      java.util.Collections.singletonMap(MetadataVersion.FEATURE_NAME,
```
Maybe this is out of scope for this PR, but it seems like the `BrokerFeatures` class just duplicates `FinalizedControllerFeatures`. Why not just use `FinalizedControllerFeatures` instead in the old controller?

The name seems `createDefault` seems misleading here -- what this is representing is the latest set of features, right?

cmccabe commented on core/src/main/scala/kafka/server/BrokerMetadataCheckpoint.scala:
```diff
@@ -102,6 +102,10 @@ class RawMetaProperties(val props: Properties = new Properties()) {
 }
 
 object MetaProperties {
+  def apply(clusterId: String, nodeId: Int): MetaProperties = {
```
why do we need this? can't we just do `new MetaProperties(clusterId, nodeId)`?

cmccabe commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -324,7 +324,8 @@ class BrokerServer(
                                                     threadNamePrefix,
                                                     config.metadataSnapshotMaxNewRecordBytes,
-                                                    metadataSnapshotter)
+                                                    metadataSnapshotter
```
it would be good to avoid this kind of whitespace change -- it makes merges harder

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/QuorumFeatures.java:
```diff
@@ -17,32 +17,73 @@
+        return features;
+    }
+
+    Optional<VersionRange> quorumSupportedFeature(String featureName) {
```
I could be wrong about this, but I think this code might be a lot simpler without streams. I think a "for" loop would make it a lot clearer what was going on

Another thing, I guess, is that null should map to 0 (if you don't have an entry in the map for a feature, you "have" that feature at level 0). So then the min is 0 (which means you can't enable it, etc.)

cmccabe commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -53,7 +55,7 @@ public class ClusterConfig {
     ClusterConfig(Type type, int brokers, int controllers, String name, boolean autoStart,
                   SecurityProtocol securityProtocol, String listenerName, File trustStoreFile,
-                  String ibp) {
+                  String ibp, MetadataVersion metadataVersion) {
```
It seems like we should parse the IBP string and deduce `MetadataVersion`, right?

mumrah replied:
Hm, actually I think I'll remove the ibp part from the annotation and just use MetadataVersion

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -927,6 +984,29 @@ private void appendRaftEvent(String name, Runnable runnable) {
+     */
+    class QuorumFeatureListener implements FeatureLevelListener {
+        @Override
+        public void handle(String featureName, short finalizedVersion) {
```
Rather than having a listener like this, why not just throw an exception from `replay()` if we can't handle the new feature level?

mumrah replied:
If we need to react to a feature level change in one of the control managers, it will probably be easier to do it via QuorumController rather than adding dependencies to FeatureControlManager. But, we don't have any use cases yet so I guess we don't really need the listener thing yet.

mumrah committed:
PR feedback (try with resources, bootstrap snapshot name)

mumrah committed:
More PR feedback (QuorumFeatures, some cleanup, more tests)

mumrah committed:
Consolidate ibp and metadataVersion in ClusterConfig for testkit

mumrah committed:
Handle the case of upgrading from KRaft preview with KIP-704

mumrah committed:
Merge remote-tracking branch 'apache-github/trunk' into KAFKA-13830-introduce-metadata-version

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -91,19 +95,11 @@ public static NodeApiVersions create(short apiKey, short minVersion, short maxVe
-                unknownApis.add(nodeApiVersion);
-            }
-        }
+    public NodeApiVersions(Collection<ApiVersion> nodeApiVersions) {
```
Is it necessary to have a special constructor for the case where there are no supported features? This won't be a common situation in the future, right? This feels like the kind of thing that could easily lead to mistakes, if someone accidentally uses it outside of test code.

Looking at my IDE I see 10 uses of this in test code. Can we just change them to use the two-argument form?

mumrah replied:
I think the constructors and factory methods on this class could use some cleanup, but it touches a lot of files so I was hoping to do this as a follow-up. For now, I'll remove the new constructor that was added

mumrah committed:
More feedback from PR

cmccabe commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -233,6 +250,19 @@ public String version() {
+    Optional<MetadataVersion> previous() {
+        int idx = this.ordinal();
+        if (idx > 1) {
+            return Optional.of(VERSIONS[idx - 2]);
```
Why is this `idx - 2` rather than `idx - 1`?

mumrah replied:
`VERSIONS` does not include the UNINITIALIZED value, so things are offset by one relative to the ordinal.

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/QuorumFeatures.java:
```diff
@@ -17,32 +17,94 @@
+    Optional<VersionRange> quorumSupportedFeature(String featureName) {
+        List<VersionRange> supportedVersions = new ArrayList<>(quorumNodeIds.size());
+        for (int nodeId : quorumNodeIds) {
+            if (nodeId == this.nodeId) {
```
What's the rationale for excluding our own `nodeId` here? I guess the idea is that it won't ever be found in `apiVersions`?

It would be good to have an INFO message if we only have version ranges from less than N peers (where N = total number of peers)

And, should we return `Optional.empty` if we only have version ranges from less than half of the peers? Or throw an exception or something? I suppose we wouldn't expect to be able to commit in this scenario, but...

mumrah replied:
Adding a log message seems reasonable here.

We _could_ add an eager check on the number of peer NodeApiVersions that we're checking, but I don't think we really need to. We add nodes to ApiVersions before any fetches can occur, and we only remove nodes from ApiVersions when processing a network disconnect. Like you said, if we have less than half of the peer's NodeApiVersions, then we wouldn't have a quorum. It also kind of seems awkward to have some quorum semantics in validation logic like this (i.e., `if W < N/2`)

cmccabe commented on metadata/src/main/java/org/apache/kafka/image/MetadataDelta.java:
```diff
@@ -253,6 +263,14 @@ public void replay(RemoveTopicRecord record) {
     public void replay(FeatureLevelRecord record) {
         getOrCreateFeaturesDelta().replay(record);
+        featuresDelta.metadataVersionChange().ifPresent(__ -> {
+            // If any feature flags change, need to trigger components to possibly handle the new flags
```
hmm, you forgot AclsDelta here :)

I think we need a separate function to handle changes in `metadataVersion`, not just `getOrCreate*Delta`.

After all, changes in `metadataVersion` could have the effect of deleting a bunch of stuff (downgrade to version X without the stuff). And this is an immediate effect: as soon as the downgrade happens, boom. The data is gone.

Whereas `getOrCreate*Delta` is a no-op if the delta object already exists.

mumrah replied:
AclsDelta ✅ 

Yea I realize there is some missing code here, but since we don't have any use downgrade cases yet I didn't want to spend too much time on this. For now, I added a no-op method on each of the Delta classes that will let them handle a metadata.version downgrade

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/util/SnapshotFileReader.java:
```diff
@@ -15,7 +15,7 @@
  */
 
-package org.apache.kafka.shell;
+package org.apache.kafka.controller.util;
```
Hmm, I don't think this is really controller-specific, is it? In fact, we don't use it in the controller per se (although it is used in the tooling...)

Perhaps `org.apache.kafka.metadata.util` ?

mumrah committed:
More PR feedback

mumrah committed:
Add handleMetadataVersionChange on Delta classes

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java:
```diff
@@ -0,0 +1,206 @@
+     * @return              The read-only bootstrap metadata
+     * @throws Exception
+     */
+    public static BootstrapMetadata load(Path bootstrapDir, MetadataVersion fallbackPreviewVersion) throws Exception {
```
hmm, don't we always treat the versions without the bootstrap file as having the MetadataVersion of 3.0? i.e. `IBP_3_0_IV0`?

mumrah replied:
There is a case with KIP-704 (leader recovery). If a preview cluster was on 3.2 and had the IBP set to this version, then if we bootstrap 3.0, we won't be able to read the newer PartitionRecord field introduced in that KIP.

cmccabe commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -231,6 +252,10 @@ object StorageTool extends Logging {
       val checkpoint = new BrokerMetadataCheckpoint(metaPropertiesPath.toFile)
       checkpoint.write(metaProperties.toProperties)
+
+      val bootstrapMetadata = BootstrapMetadata.create(metadataVersion)
```
should we write this first, so that if it fails, we don't have a valid directory?

mumrah replied:
The bootstrap file and meta.properties are in the same directory, so L254 should cover that

cmccabe replied:
ok

mumrah committed:
Add logging when we don't have a full quorum

mumrah committed:
If IBP is not configured, use 3.0 in bootstrap metadata

cmccabe commented:
LGTM
